### DISCUSSION TITLE
multi pass box blur compute shader

### DISCUSSION THREAD

**Discussion by martin19** [gpuweb/gpuweb#2509] [Q&A]
If you have a Q&A style question about using WebGPU, consider using a GitHub "Discussion".Hi, 
I'm working on a multipass compute shader to compute a blurred image.
I have 3 textures, 1. for input, 2.&3. for ping pong the previous to the next result.
In the implementation I need to compute 3 passes of horizontal box blur and 3 
passes of vertical box blur to approximate gaussian blur.
(similar to "Moving averages compute shader version" @ https://www.intel.com/content/www/us/en/developer/articles/technical/an-investigation-of-fast-real-time-gpu-based-image-blur-algorithms.html)

```
let computePass = commandEncoder.beginComputePass();
computePass.setPipeline(blurPipeline);
device.queue.writeBuffer(blurParamsBuffer, 0, new Uint32Array([settings.filterSize, 1])); //<-- 1 tells the shader to execute horizontal pass
computePass.setBindGroup(0, computeBindGroup0); // <-- compute 1. pass horizontal (texInput->tex0)
computePass.dispatch(32,1,1);
computePass.setBindGroup(0, computeBindGroup1); // <-- compute 2. pass horizontal (tex0->tex1)
computePass.dispatch(32,1,1);
computePass.setBindGroup(0, computeBindGroup2); // <-- compute 3. pass horizontal (tex1->tex0)
computePass.dispatch(32,1,1);
device.queue.writeBuffer(blurParamsBuffer, 0, new Uint32Array([settings.filterSize, 0])); //<-- 0 tells the shader to execute vertical pass
computePass2.setBindGroup(0, computeBindGroup1); // <-- compute 1. pass vertical (tex0->tex1)
computePass2.dispatch(32,1,1);
computePass2.setBindGroup(0, computeBindGroup2); // <-- compute 2. pass vertical (tex1->tex0)
computePass2.dispatch(32,1,1);
computePass2.setBindGroup(0, computeBindGroup1); // <-- compute 3. pass vertical (tex0->tex1)
computePass2.dispatch(32,1,1);
computePass.endPass();
//display tex1
```

Questions: 
- The result displayed seems to be after 3. vertical pass, horizontal passes are *not* executed
- is this setup legit? I'm setting radius and direction (horizontal, vertical) of the blur in 
blurParamsBuffer inside the computePass.
- how can I make sure the workloads (dispatch(32,1,1)) are all executed in sequence

kind regards, Martin

**Comment by Kangz** [comment]
Please follow the instructions from the issue wizard, both this issue and #2508 were filed ignoring the instructions in the wizard. I'll convert the issue to a discussion and will answer.

**Reply by Kangz** [reply]
The problem here is because you use `writeBuffer` twice without a queue submission in the middle. `writeBuffer` is associated with the queue timeline, while commands in a command encoder (and a compute pass encoder) are added to the queue timeline only on submit. So what happens on the queue in your example is the following, in this order:

 - First `writeBuffer`, for horizontal blur.
 - Second `writeBuffer`, for vertical blur.
 - `submit` that will only see the data from the second `writeBuffer` and only perform vertical blurs.

There's several options for what you could do here:

 - You could have two uniform buffers: one for the horizontal blur and one for the vertical blur.
 - You could do a `copyBufferToBuffer` when switching from horizontal to vertical. Since the `copyBufferToBuffer` command is in the command encoder, it will execute in the correct order with respect to the compute passes.

Of note, all commands in WebGPU execute in the order they are specified in the queue and all synchronization is done for you by the implementation. (there's an exception for draws in a render pass but that doesn't matter for your case).

**Comment by martin19** [comment]
Hi thanks for your explanations

i had some success with copyBufferToBuffer.
I'm now using the same pipeline 6 times and use copyBufferToBuffer to pass different radii/direction parameters to the shader.

the compute blocks look like so:

```
commandEncoder.copyBufferToBuffer(blurParams[1], 0, blurParamsBuffer, 0, 8);
computePass = commandEncoder.beginComputePass();
computePass.setPipeline(blurPipeline);
computePass.setBindGroup(0, computeBindGroup1);
computePass.dispatch(workgroups,1,1);
computePass.endPass();
```

The result is there and looks correct, so far so good.

However, performace wise I expected this approch to be about 10x faster, even on my low spec UHD620.
Even the non-parallel plain javascript version of this algorithm outperforms this shader, there must be something wrong.
Tried different parameters for workgroup (8,16,32,64,128) - no luck.

The shader code is here:
https://gist.github.com/martin19/1a9b60e511d21f59d2132163244c110e

The webgpu code is here:
https://gist.github.com/martin19/984f2c452b5b50be30a80373c9be915f

Do you have any hints if I'm doing something completely wrong?

**Reply by Kangz** [reply]
Here you are making your compute shaders loop over the whole image, so you can image like if you are launching a dozen core of the GPU to do the whole compute operation, instead of the ~100 cores your GPU has. It's better to make your workload computable in independent units that can complete in parallel.

Also workgroups seems to be used both for the workgroup size and the number of workgroups to launch? That can't be right. 

As a side note you could use a single uniform buffer and great bindgroups at an offset in it. This would be a bit more efficient (though for 6 buffers it's a wash).

**Reply by martin19** [reply]
seems I misunderstood a few things, I thought each compute pass processes #workgroup stripes of (imageheight/#workgroup).
Will have a look, thanks!

**Comment by martin19** [comment]
Setting "workgroup_size(1, 1, 1)" and workgroups in dispatch to 128 gives pretty acceptable result (full-hd blur works now at "interactive" rate) . I still don't understand what this parameter really does. 
Can you please explain - that spec is a tough read :)

The idea of the parallelization of the algorithm is to divide the image into N stripes and process each stripe on a GPU core.
There are 6 passes total - 3xhorizontal boxblur, 3xvertical boxblur. The 6 passes need to be called sequentially since each step uses the result of the step before. I don't see any other possibility to further parallelize this algorithm. The pro of this algorithm is for large blur radii only, that is what I'm looking for.

Thanks again for your help!

**Reply by Kangz** [reply]
Please look at https://web.dev/gpu-compute/ I think it explains it a bit, or search for compute shader tutorials. `workgroup_size` of 1, 1, 1 is really bad because out of your GPU's 8 ALUs in a workgroup a single one will be used.

There are also matrix channels listed in https://github.com/gpuweb/gpuweb/wiki/Contact where the community hangs out and more people could discuss things "live".
